### PR TITLE
Add NavigationBuilder<TSource, TTarget>

### LINKED ISSUE #21942: API consistency test to ensure generic Fluent API classes hide inherited non-generic methods

### LINKED ISSUE #22012: Add NavigationBuilder<TSource, TTarget>

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Add API consistency test to ensure generic Fluent API classes hide inherited non-generic methods

Fixes #22012
Fixes #21942

### CODE REVIEW

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs [hunk 1]
```diff
@@ -347,35 +347,6 @@ public virtual void Skip_navigation_field_can_be_set_via_attribute()
                 Assert.Equal("_randomField", model.FindEntityType(typeof(ManyToManyNavPrincipal)).FindSkipNavigation("Dependents").GetFieldName());
             }
 
-            [ConditionalFact]
-            public virtual void Navigation_properties_can_set_access_mode_using_navigation_names()
```
Comment (ajcvickers) [inline test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs:351]: Why remove this test?

Comment (ajcvickers) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs:351]: Never mind; I see.

File: test/EFCore.Tests/ApiConsistencyTestBase.cs [hunk 2]
```diff
@@ -907,6 +986,27 @@ protected ApiConsistencyFixtureBase()
             public abstract bool TryGetProviderOptionsDelegate (out Action<DbContextOptionsBuilder> configureOptions);
 
             public virtual HashSet<Type> FluentApiTypes { get; } = new HashSet<Type>();
+
+            public virtual Dictionary<Type, Type> GenericFluentApiTypes { get; } = new Dictionary<Type, Type>
```
Comment (ajcvickers) [inline test/EFCore.Tests/ApiConsistencyTestBase.cs:990]: So when we create a new builder we need to remember to add it to this list?

File: test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs (hunk 1 above)
Comment (AndriySvyryd) [reply test/EFCore.Tests/ModelBuilding/ManyToManyTestBase.cs:351]: It's covered by the expression-based test being run on the non-generic implementation

File: test/EFCore.Tests/ApiConsistencyTestBase.cs (hunk 2 above)
Comment (AndriySvyryd) [reply test/EFCore.Tests/ApiConsistencyTestBase.cs:990]: Yes

File: src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs [hunk 3]
```diff
@@ -167,8 +167,26 @@ public virtual PropertyBuilder<TProperty> Property<TProperty>([NotNull] Expressi
...
+        /// <summary>
+        ///         Returns an object that can be used to configure an existing navigation property of the entity type.
+        ///         It is an error for the navigation property not to exist.
+        /// </summary>
+        /// <param name="navigationExpression">
+        ///     A lambda expression representing the navigation property to be configured (
+        ///     <c>blog => blog.Posts</c>).
+        /// </param>
+        /// <returns> An object that can be used to configure the navigation property. </returns>
+        public virtual NavigationBuilder<TEntity, TNavigation> Navigation<TNavigation>(
```
Comment (smitpatel) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder`.cs:186]: typeparam in XML docs?

**Pull request merged**